src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -78,8 +78,136 @@ import java.util.*
  *      ```
  */
 
+class TransactionalBankAccount(accountNumber: String,
+                               accountHolderName: String,
+                               balance: Double = 0.0
+) : BankAccount(accountNumber, accountHolderName, balance) {
 
+    private val transactions = mutableListOf<Transaction>()
 
+    override fun deposit(amount: Double): Boolean {
+        val success = super.deposit(amount)
+        recordTransaction(amount, success, TransactionType.DEPOSIT)
+        return success
+    }
+
+    override fun withdraw(amount: Double): Boolean {
+        val success = super.withdraw(amount)
+        recordTransaction(amount, success, TransactionType.WITHDRAWAL)
+        return success
+    }
+
+    private fun recordTransaction(amount: Double, success: Boolean, transactionType: TransactionType) {
+        val transaction = Transaction(
+            LocalDateTime.now(),
+            transactionType,
+            amount,
+            if (transactionType == TransactionType.DEPOSIT) {
+                if (amount > 0)
+                    getBalance() + amount
+                else {
+                    getBalance()
+                }
+            } else {
+                getBalance() - amount
+            },
```

Comment:
Why not simply read the balance before and after the transaction?

Reply:
It was an early solution to the problem, but it felt like the code was repetitive, so I tried implementing it like this instead. However, since you recommended the previously tried approach and since the logic is faulty according to your following comment, I will implement it that way.

Reply:
I don't mind removing a duplication, the problem with a current approach is that it's not correct. :) 

I would simply refactor your solution to having something like: 
```
fun doTransaction(amount: Int, transactionType: TransactionType): TransactionStatus
```
And that can be reused in both methods.

Reply:
Yes, I understand now what you meant. Thank you! :)